swarm/storage/resource.go

```diff
@@ -0,0 +1,529 @@
+	copy(chunk.SData[16:], val)
+	self.Put(chunk)
+	log.Debug("new resource", "name", validname, "key", ensname, "startblock", currentblock, "frequency", frequency)
+
+	self.resourceLock.Lock()
+	defer self.resourceLock.Unlock()
+	self.resources[name] = &resource{
+		name:       validname,
+		ensname:    ensname,
+		startblock: currentblock,
+		frequency:  frequency,
+		updated:    time.Now(),
+	}
+	return self.resources[name], nil
+}
+
+// Set an externally defined resource object
+//
+// If the resource update root chunk is located externally (for example as a normal
+// chunk looked up by ENS) the data would be manually added with this method).
+//
+// Method will fail if resource is already registered in this session, unless
+// `allowOverwrite` is set
+func (self *ResourceHandler) SetResource(rsrc *resource, allowOverwrite bool) error {
+
+	if rsrc.name == "" {
+		return fmt.Errorf("Resource name cannot be empty")
+	}
+
+	utfname, err := idna.ToUnicode(rsrc.name)
+	if err != nil {
+		return fmt.Errorf("Invalid IDNA rsrc name '%s'", rsrc.name)
+	}
+	if !allowOverwrite {
+		if _, ok := self.resources[utfname]; ok {
+			return fmt.Errorf("Resource exists")
+		}
+	}
+
+	// get our blockheight at this time
+	currentblock, err := self.getBlock()
+	if err != nil {
+		return err
+	}
+
+	if rsrc.startblock > currentblock {
+		return fmt.Errorf("Startblock cannot be higher than current block (%d > %d)", rsrc.startblock, currentblock)
+	}
+
+	if rsrc.frequency == 0 {
+		return fmt.Errorf("Frequency cannot be 0")
+	}
+
+	if len(rsrc.ensname) > 0 {
+		ensname := ens.EnsNode(rsrc.name)
+		if ensname != rsrc.ensname {
+			return fmt.Errorf("Namehash %x is not a valid namehash of IDNA name '%s'", rsrc.ensname, rsrc.name)
+		}
+	}
+	self.resources[utfname] = rsrc
+	return nil
+}
+
+// Searches and retrieves the last version of the resource update identified by `name`
+//
+// It starts at the next period after the current block height, and upon failure
+// tries the corresponding keys of each previous period until one is found
+// (or startblock is reached, in which case there are no updates).
+// If an update is found, version numbers are iterated until failure, and the last
+// successfully retrieved version is copied to the corresponding resources map entry
+// and returned.
+//
+// If refresh is set to true, the resource data will be reloaded from the resource update
+// root chunk.
+// It is the callers responsibility to make sure that this chunk exists (if the resource
+// update root data was retrieved externally, it typically doesn't)
+func (self *ResourceHandler) OpenResource(name string, refresh bool) (*resource, error) {
+
+	rsrc := &resource{}
+
+	// if the resource is not known to this session we must load it
+	// if refresh is set, we force load
+	if _, ok := self.resources[name]; !ok || refresh {
+
+		// make sure our ens identifier is idna safe
+		validname, err := idna.ToASCII(name)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
Another example for https://github.com/ethersphere/go-ethereum/pull/139/files#r156192771, let's reduce duplicated code